src/groups/mqb/mqbstat/mqbstat_jsonprinter.cpp

```diff
@@ -0,0 +1,268 @@
+                             const mwcst::StatContext& ctx) const;
+    void populateQueueStats(bdljsn::JsonObject*       domainObject,
+                            const mwcst::StatContext& ctx) const;
+    void populateDomainQueueStats(bdljsn::JsonObject*       parent,
+                                  const mwcst::StatContext& ctx) const;
+
+  public:
+    // TRAITS
+    BSLMF_NESTED_TRAIT_DECLARATION(JsonPrinterImpl, bslma::UsesBslmaAllocator)
+
+    // CREATORS
+
+    /// Create a new `JsonPrinterImpl` object, using the specified `config`,
+    /// `statContextsMap` and the specified `allocator`.
+    JsonPrinterImpl(const mqbcfg::StatsConfig& config,
+                    const StatContextsMap&     statContextsMap,
+                    bslma::Allocator*          allocator);
+
+    // ACCESSORS
+
+    /// Print the json-encoded stats to the specified `out`.
+    /// If the specified `compact` flag is `true`, the json is printed in a
+    /// compact form, otherwise the json is printed in a pretty form.
+    /// Return `0` on success, and non-zero return code on failure.
+    ///
+    /// THREAD: This method is called in the *StatController scheduler* thread.
+    int printStats(bsl::string* out, bool compact) const;
+};
+
+inline JsonPrinterImpl::JsonPrinterImpl(const mqbcfg::StatsConfig& config,
+                                        const StatContextsMap& statContextsMap,
+                                        bslma::Allocator*      allocator)
+: d_config(config)
+, d_opsCompact(bdljsn::WriteOptions().setSpacesPerLevel(0).setStyle(
+      bdljsn::WriteStyle::e_COMPACT))
+, d_opsPretty(bdljsn::WriteOptions().setSpacesPerLevel(4).setStyle(
+      bdljsn::WriteStyle::e_PRETTY))
+, d_contexts(statContextsMap, allocator)
+{
+}
+
+inline void
+JsonPrinterImpl::populateQueueValues(bdljsn::JsonObject*       queueObject,
+                                     const mwcst::StatContext& ctx) const
+{
+    // PRECONDITIONS
+    BSLS_ASSERT_SAFE(queueObject);
+
+    static const bsl::vector<Stat::Enum> defs = {Stat::e_NB_PRODUCER,
+                                                 Stat::e_NB_CONSUMER,
+                                                 Stat::e_PUT_MESSAGES_DELTA,
+                                                 Stat::e_PUT_BYTES_DELTA,
+                                                 Stat::e_PUSH_MESSAGES_DELTA,
+                                                 Stat::e_PUSH_BYTES_DELTA,
+                                                 Stat::e_ACK_DELTA,
+                                                 Stat::e_ACK_TIME_AVG,
+                                                 Stat::e_ACK_TIME_MAX,
+                                                 Stat::e_NACK_DELTA,
+                                                 Stat::e_CONFIRM_DELTA,
+                                                 Stat::e_CONFIRM_TIME_AVG,
+                                                 Stat::e_CONFIRM_TIME_MAX};
```

Comment:
Todo: safe static variable initialization in a multithreaded env. Cache as a const class field instead.